### ISSUE TITLE
Memory leak - Memoize not being garbage collected

### ISSUE THREAD

**Issue by AbdealiLoKo** [AbdealiJK/file-metadata#43]
The `fetch` function being used in the class is memoized to make it faster and avoid repetitive read/writing and so on. It seems that the memoization is never cleared, because the dictionary where the cache is stored never goes out of reference.

What needs to be done is to figure out a method to garbage collect this memoized information correctly.

Due to this, the bulk analysis bot cannot run more than ~1500 files at a time in ToolsLab. The memoy becomes full and it collapses.

**Comment by jayvdb** [comment]
Try using an existing library to do this
https://pypi.python.org/pypi?%3Aaction=search&term=memoized&submit=search

Also there are requests cachers, like cachecontrol. https://github.com/ionrock/cachecontrol

**Comment by drtrigon** [comment]
From what I read they might only be caching IF the webpage allows it. While this might be true for commons and thus the bot - it might not always be true for other sources.

Am 19. Juni 2016 01:44:30 MESZ, schrieb John Vandenberg notifications@github.com:

> Try using an existing library to do this
> https://pypi.python.org/pypi?%3Aaction=search&term=memoized&submit=search
> 
> Also there are requests cachers, like cachecontrol.
> https://github.com/ionrock/cachecontrol
> 
> ---
> 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/AbdealiJK/file-metadata/issues/43#issuecomment-226971135

Dr. Trigon

**Comment by AbdealiLoKo** [comment]
The `requests` cache control was not what I was looking for. I was cacheing the NDarray with a memoized function. I tried a lot of the popular `memoize` packages in PyPI, and all of them were intended for global functions, not class methods.

This is now fixed in https://github.com/AbdealiJK/file-metadata/commit/bcb071907a40fe811f43f7798d1064817b08a39f

Tl;DR:
The issue with that approach is that the first argument (`self`) is used in the Key of the cache. Which implies that the object is being referenced. This makes the python Garbage collector to think that the object is being used and it does not delete it.

I found a snippet of a nice class method memoizer at http://code.activestate.com/recipes/577452-a-memoize-decorator-for-instance-methods/
I implemented it and tested it by running it on 3000 files and the VIRT memory from `top` was less than 1.5GB at all times. (Earlier it went more than 2GB in less than 200 files)

It works by saving the cache (dict) in `<object>.__cache`. It uses the first argument given to the function as the class. Hence, when the object is garbage collected, the cache is garbage collected too.

**Comment by drtrigon** [comment]
Nice!

Am 20. Juni 2016 08:28:08 MESZ, schrieb AbdealiJK notifications@github.com:

> The `requests` cache control was not what I was looking for. I was
> cacheing the NDarray with a memoized function. I tried a lot of the
> popular `memoize` packages in PyPI, and all of them were intended for
> global functions, not class methods.
> 
> This is now fixed in
> https://github.com/AbdealiJK/file-metadata/commit/bcb071907a40fe811f43f7798d1064817b08a39f
> 
> Tl;DR:
> The issue with that approach is that the first argument (`self`) is
> used in the Key of the cache. Which implies that the object is being
> referenced. This makes the python Garbage collector to think that the
> object is being used and it does not delete it.
> 
> I found a snippet of a nice class method memoizer at
> http://code.activestate.com/recipes/577452-a-memoize-decorator-for-instance-methods/
> I implemented it and tested it by running it on 3000 files and the VIRT
> memory from `top` was less than 1.5GB at all times. (Earlier it went
> more than 2GB in less than 200 files)
> 
> It works by saving the cache (dict) in `<object>.__cache`. It uses the
> first argument given to the function as the class. Hence, when the
> object is garbage collected, the cache is garbage collected too.
> 
> ---
> 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/AbdealiJK/file-metadata/issues/43#issuecomment-227059765

Dr. Trigon
